File: isegoria/src/main/java/com/eulersbridge/isegoria/election/efficacy/SelfEfficacyQuestionsFragment.java
package com.eulersbridge.isegoria.election.efficacy;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import com.eulersbridge.isegoria.MainActivity;
import com.eulersbridge.isegoria.R;
import com.eulersbridge.isegoria.util.ui.TitledFragment;

public class SelfEfficacyQuestionsFragment extends Fragment implements TitledFragment, MainActivity.TabbedFragment {

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.self_efficacy_questions_fragment, container, false);

        SelfEfficacySliderBar sliderBar1 = rootView.findViewById(R.id.selfEfficacySliderBar1);
        SelfEfficacySliderBar sliderBar2 = rootView.findViewById(R.id.selfEfficacySliderBar2);
        SelfEfficacySliderBar sliderBar3 = rootView.findViewById(R.id.selfEfficacySliderBar3);
        SelfEfficacySliderBar sliderBar4 = rootView.findViewById(R.id.selfEfficacySliderBar4);

        EfficacyQuestionsViewModel viewModel = ViewModelProviders.of(this).get(EfficacyQuestionsViewModel.class);

        viewModel.score1.observe(this, score1 -> {
            if (score1 != null)
                sliderBar1.setScore(score1);
        });
        viewModel.score2.observe(this, score2 -> {
            if (score2 != null)
                sliderBar2.setScore(score2);
        });
        viewModel.score3.observe(this, score3 -> {
            if (score3 != null)
                sliderBar3.setScore(score3);
        });
        viewModel.score4.observe(this, score4 -> {
            if (score4 != null)
                sliderBar4.setScore(score4);
        });

        Button doneButton = rootView.findViewById(R.id.selfEfficacyDoneButton);
        doneButton.setOnClickListener(view -> {
            view.setEnabled(false);

            viewModel.addUserEfficacy().observe(this, success -> {
                if (success != null && success) {

                    if (getActivity() != null)
                        getActivity().getSupportFragmentManager().popBackStack();

                } else {
                    view.setEnabled(true);
                }
            });
        });

        return rootView;
    }

    @Override
    public String getTitle(Context context) {
        return context.getString(R.string.section_title_self_efficacy_questions);
    }

    @Override
    public void setupTabLayout(TabLayout tabLayout) {
        tabLayout.setVisibility(View.GONE);
    }
}

File: isegoria/src/main/kotlin/com/eulersbridge/isegoria/election/efficacy/SelfEfficacyQuestionsFragment.kt
package com.eulersbridge.isegoria.election.efficacy

import android.arch.lifecycle.ViewModelProvider
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.eulersbridge.isegoria.R
import com.eulersbridge.isegoria.util.extension.observe
import com.eulersbridge.isegoria.util.extension.observeBoolean
import dagger.android.support.AndroidSupportInjection
import kotlinx.android.synthetic.main.self_efficacy_questions_fragment.*
import javax.inject.Inject

// TODO: Convert into Activity, as this fragment is an app-level modal
class SelfEfficacyQuestionsFragment : Fragment() {

    @Inject
    internal lateinit var modelFactory: ViewModelProvider.Factory
    private lateinit var viewModel: EfficacyViewModel

    override fun onAttach(context: Context?) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
        viewModel = ViewModelProviders.of(this, modelFactory)[EfficacyViewModel::class.java]
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View?
        = inflater.inflate(R.layout.self_efficacy_questions_fragment, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val pairs = mapOf(
                viewModel.score1 to sliderBar1,
                viewModel.score2 to sliderBar2,
                viewModel.score3 to sliderBar3,
                viewModel.score4 to sliderBar4
        )

        for ((liveData, sliderBar) in pairs) {
            observe(liveData) {
                if (it != null)
                    sliderBar.score = it.toInt()
            }
        }

        doneButton.setOnClickListener {
            viewModel.score1.value = sliderBar1.score
            viewModel.score2.value = sliderBar2.score
            viewModel.score3.value = sliderBar3.score
            viewModel.score4.value = sliderBar4.score
            viewModel.onDone()
        }

        createViewModelObservers()
    }

    private fun createViewModelObservers() {
        observeBoolean(viewModel.doneButtonEnabled) {
            doneButton.isEnabled = it
        }

        observe(viewModel.efficacyComplete) {
            activity!!.supportFragmentManager.popBackStack()
        }
    }
}
